Flaky-test: org.apache.pulsar.client.api.MemoryLimitTest#testRejectMessages

https://github.com/apache/pulsar/runs/5138925372?check_suite_focus=true
```
Error:  Tests run: 8, Failures: 2, Errors: 0, Skipped: 6, Time elapsed: 11.631 s <<< FAILURE! - in org.apache.pulsar.client.api.MemoryLimitTest
Error:  testRejectMessages(org.apache.pulsar.client.api.MemoryLimitTest)  Time elapsed: 2.115 s  <<< FAILURE!
java.lang.AssertionError: expected [103424] but found [62464]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:797)
	at org.testng.Assert.assertEquals(Assert.java:807)
	at org.apache.pulsar.client.api.MemoryLimitTest.testRejectMessages(MemoryLimitTest.java:86)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

--- a/pulsar-broker/src/test/java/org/apache/pulsar/client/api/MemoryLimitTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/client/api/MemoryLimitTest.java
@@ -18,19 +18,21 @@
  */
 package org.apache.pulsar.client.api;
 
-import static org.testng.Assert.assertEquals;
-import static org.testng.Assert.fail;
-
-import java.util.concurrent.CountDownLatch;
-
 import lombok.Cleanup;
-
 import org.apache.pulsar.client.api.PulsarClientException.MemoryBufferIsFullError;
-import org.apache.pulsar.client.impl.PulsarClientImpl;
+import org.apache.pulsar.client.impl.ProducerImpl;
+import org.apache.pulsar.client.impl.PulsarTestClient;
+import org.awaitility.Awaitility;
 import org.testng.annotations.AfterMethod;
 import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
 
 @Test(groups = "broker-api")
 public class MemoryLimitTest extends ProducerConsumerBase {
@@ -62,27 +64,30 @@ public class MemoryLimitTest extends ProducerConsumerBase {
             throws Exception {
         String topic = newTopicName();
 
-        @Cleanup
-        PulsarClientImpl client = (PulsarClientImpl) PulsarClient.builder()
+        ClientBuilder clientBuilder = PulsarClient.builder()
                 .serviceUrl(pulsar.getBrokerServiceUrl())
-                .memoryLimit(100, SizeUnit.KILO_BYTES)
-                .build();
+                .memoryLimit(100, SizeUnit.KILO_BYTES);
 
         @Cleanup
-        Producer<byte[]> producer = client.newProducer()
+        PulsarTestClient client = PulsarTestClient.create(clientBuilder);
+
+        @Cleanup
+        ProducerImpl<byte[]> producer = (ProducerImpl<byte[]>) client.newProducer()
                 .topic(topic)
                 .blockIfQueueFull(false)
+                .sendTimeout(5, TimeUnit.SECONDS)
                 .create();
 
+        // make sure all message pending at pendingMessages queue
+        // connection with broker can not be established, so handleSendReceipt will not be invoked while sending message
+        client.dropOpSendMessages();
         final int n = 101;
-        CountDownLatch latch = new CountDownLatch(n);
-
         for (int i = 0; i < n; i++) {
-            producer.sendAsync(new byte[1024]).thenRun(() -> {
-                latch.countDown();
-            });
+            producer.sendAsync(new byte[1024]);
         }
-
+        Awaitility.await()
+                .atMost(Duration.ofSeconds(5))
+                .until(() -> producer.getPendingQueueSize() == n);
         assertEquals(client.getMemoryLimitController().currentUsage(), n * 1024);
 
         try {
@@ -92,8 +97,10 @@ public class MemoryLimitTest extends ProducerConsumerBase {
             // Expected
         }
 
-        latch.await();
-
+        client.allowReconnecting();
+        Awaitility.await()
+                .atMost(Duration.ofSeconds(30))
+                .until(() -> producer.getPendingQueueSize() == 0);
         assertEquals(client.getMemoryLimitController().currentUsage(), 0);
 
         // We should now be able to send again
@@ -105,41 +112,40 @@ public class MemoryLimitTest extends ProducerConsumerBase {
         String t1 = newTopicName();
         String t2 = newTopicName();
 
-        @Cleanup
-        PulsarClientImpl client = (PulsarClientImpl) PulsarClient.builder()
+        ClientBuilder clientBuilder = PulsarClient.builder()
                 .serviceUrl(pulsar.getBrokerServiceUrl())
-                .memoryLimit(100, SizeUnit.KILO_BYTES)
-                .build();
+                .memoryLimit(100, SizeUnit.KILO_BYTES);
 
         @Cleanup
-        Producer<byte[]> p1 = client.newProducer()
+        PulsarTestClient client = PulsarTestClient.create(clientBuilder);
+
+        @Cleanup
+        ProducerImpl<byte[]> p1 = (ProducerImpl<byte[]>) client.newProducer()
                 .topic(t1)
                 .blockIfQueueFull(false)
+                .sendTimeout(5, TimeUnit.SECONDS)
                 .create();
 
         @Cleanup
-        Producer<byte[]> p2 = client.newProducer()
+        ProducerImpl<byte[]> p2 = (ProducerImpl<byte[]>) client.newProducer()
                 .topic(t2)
                 .blockIfQueueFull(false)
+                .sendTimeout(5, TimeUnit.SECONDS)
                 .create();
 
+        client.dropOpSendMessages();
         final int n = 101;
-        CountDownLatch latch = new CountDownLatch(n);
-
         for (int i = 0; i < n / 2; i++) {
-            p1.sendAsync(new byte[1024]).thenRun(() -> {
-                latch.countDown();
-            });
-            p2.sendAsync(new byte[1024]).thenRun(() -> {
-                latch.countDown();
-            });
+            p1.sendAsync(new byte[1024]);
+            p2.sendAsync(new byte[1024]);
         }
 
         // Last message in order to reach the limit
-        p1.sendAsync(new byte[1024]).thenRun(() -> {
-            latch.countDown();
-        });
+        p1.sendAsync(new byte[1024]);
 
+        Awaitility.await()
+                .atMost(Duration.ofSeconds(5))
+                .until(() -> (p1.getPendingQueueSize() + p2.getPendingQueueSize()) == n);
         assertEquals(client.getMemoryLimitController().currentUsage(), n * 1024);
 
         try {
@@ -156,8 +162,10 @@ public class MemoryLimitTest extends ProducerConsumerBase {
             // Expected
         }
 
-        latch.await();
-
+        client.allowReconnecting();
+        Awaitility.await()
+                .atMost(Duration.ofSeconds(30))
+                .until(() -> (p1.getPendingQueueSize() + p2.getPendingQueueSize()) == 0);
         assertEquals(client.getMemoryLimitController().currentUsage(), 0);
 
         // We should now be able to send again

